integrate oxidation state prediction into CP2K workchains

ltalirz opened the pull request:
It is now possible to specify the `initial_magetization` in cp2k workchains in the following ways:

 * `"zero"` for no initial magnetization
 *  `"element"` for a guess based on the element
 * `"oxidation_state"`  for guess based on the oxidation state predicted for the neighborhood of the metal using the oximachine (corresponding to maximally unpaired spins)

Users can also provide user-defined values in the form of a dictionary element=>magnetization or a dictionary in the form of the `initial_magnetizations.yaml` file.

ltalirz committed:
tag kinds depending on oxidation state

ltalirz commented:
@danieleongari In addition to the "tagged structuredata" that is provided here, 
you may also want a dictionary kind -> oxidation state.

I'm happy to make the changes if you let me know where the changes should go - 

 1. should there be a switch for using oximachine in the protocol.yaml (I guess yes?)?
 2. in which protocols should it be "on"? should there be a new one?
 3. i suggest the oxidation states will then be computed once in `setup_multistage` and then passed around to be used in `get_input_multiplicity` and `get_kinds_section`

danieleongari commented:
Hi, thanks for your work!

0. Here we need to provide some conversion table: `kind_magnetization(element, oxidation_state)`. I suggest to make it as a .csv file in the `aiida_lsmo/workchains/cp2k_multistage_protocols/` so that later we can use different conventions, since we are adding and extra degree of freedom.
1. Yes, instead of using elemennts, the user needs to specify in `protocol.yaml` something like: `initial_magnetization: { logic: highest_magnetization }. Let's not complicate stuff: one can or specify all the atoms or use the logic. If the user wants to specify a magetization he specifies all the atoms. LEt's also introduce the logic: if no logic specified and no element specified, that element witll have magnetization 0.
2. At the moment, I'm using specific yaml protocols loaded as input: we can test this way and later decide how to set defaults.
3. Ok, I will consider your suggestion when introducing `kind_magnetization(element, oxidation_state)`.

ltalirz committed:
use oxidation state in cp2k work chains

and incorporate suggestions from code review

ltalirz commented:
> One last concern: we need to be careful about oximachine running locally when hundreds/thousands or jobs are submitted!

Yes... however, I'm not sure it make sense to move it somewhere else

ltalirz commented:
@danieleongari I've incorporated your table into the workchain now.

 * I've converted it to yaml so that one can add comments etc.
 * I've added a "default oxidation state" (from the old magnetization input) so that the table can be used in both cases
 * The method of choosing the initial magnetization is now simply a string in the protocol.yaml. If you prefer to be able to do either string or manual table, I can reintroduce it... 

Unfortunately, there is a bit of logic needed to do the "tagging" + the mapping to magnetization states; in order to avoid having too many function calls and variables passed around I created a class for it.

It's still not very elegant; one can probably still improve it.

ltalirz commented:
Ah hm, I see that in `examples/test_multistage_aluminum.py` the `initial_magnetization` is actually specified manually (and only for `Al`).

I can reintroduce this as a possibility... just a bit more logic

danieleongari commented on aiida_lsmo/workchains/cp2k_multistage_protocols/__init__.py:
```diff
@@ -8,6 +8,25 @@
 PROTOCOL_DIR = Path(__file__).resolve().parent
 
+with open(PROTOCOL_DIR / 'initial_magnetization.yaml') as handle:
+    INITIAL_MAGNETIZATION = yaml.load(handle)
```
`yaml.safe_load()` not to have:
```
In most other cases you should consider using 'safe_load(stream)'
  INITIAL_MAGNETIZATION = yaml.load(handle)
```

danieleongari commented on aiida_lsmo/utils/cp2k_utils.py:
```diff
@@ -1,54 +1,151 @@
+                })
+        else:
+            kinds = [{
+                '_': k['element'],
```
This is not correctly printing the kind. 
From some testing for Cu(I)/Cu(II)-HKUST-1 I get:
```
      &KIND Cu
         BASIS_SET DZVP-MOLOPT-SR-GTH-q11
         MAGNETIZATION 0
         POTENTIAL GTH-PBE-q11
      &END KIND
      &KIND Cu
         BASIS_SET DZVP-MOLOPT-SR-GTH-q11
         MAGNETIZATION 1
         POTENTIAL GTH-PBE-q11
      &END KIND
```
Just as a sinde note: CP2K should strip the element, from e.g., CuX, if I remember well.
But to be sure one can add the field `ELEMENT`.
https://manual.cp2k.org/trunk/CP2K_INPUT/FORCE_EVAL/SUBSYS/KIND.html#list_ELEMENT

danieleongari commented on tests/test_oxidation_state.py:
```diff
@@ -23,3 +25,30 @@ def test_oxidation_state(cif_filename):
+    structure = StructureData(ase=atoms)
+
+    subsys = Cp2kSubsys(structure=structure,
+                        oxidation_states=OXIMACHINE_RUNNER.run_oximachine(atoms),
```
Since it is taking some time, maybe print in the `report` that the oximachine is running

ltalirz replied:
thanks, done

ltalirz replied:
fixed

danieleongari commented:
Thanks, it is now working great.
I used this script as example, if you want to adapt it as a test:
```
import ase

from aiida.plugins import DataFactory, WorkflowFactory
from aiida import cmdline
from aiida import engine
from aiida.orm import Dict, StructureData, Str, Int, SinglefileData, Code

# Workchain objects
Cp2kMultistageWorkChain = WorkflowFactory('lsmo.cp2k_multistage')

# Data objects
StructureData = DataFactory('structure')


# testing user change of parameters and protocol
parameters = Dict(dict={'FORCE_EVAL': {'DFT': {'MGRID': {'CUTOFF': 250,}}}})

# Construct process builder
builder = Cp2kMultistageWorkChain.get_builder()
builder.structure = StructureData(ase=ase.io.read('../tests/data/Cu-I-II-HKUST-1.cif'))
builder.protocol_tag = Str('standard')
builder.cp2k_base.cp2k.parameters = parameters
builder.cp2k_base.cp2k.code = Code.get_from_string('cp2k-5.1@helvetios')
builder.cp2k_base.cp2k.metadata.options.resources = {
    'num_machines': 1,
 #   'num_mpiprocs_per_machine': 32,
}
builder.cp2k_base.cp2k.metadata.options.max_wallclock_seconds = 1 * 60 * 60

wc = engine.submit(builder)

```

ltalirz committed:
generalize 'initial_magnetization' field

allow user to specify initial magnetizations

ltalirz committed:
add cu-hkust example

also:
 * make protocol schema less strict
 * fix: cp2k multistage ghost section

ltalirz commented:
the last failing test is because something changed in the cp2k input of the binding energy workchain (and I would need to regenerate the test data).
this leads to the general question, for which protocols the oxidation state prediction should be the default

we now have three pre-defined methods:
"zero" (no initial mags), "element" (using default oxidation state) and "oxidation_state" (using predicted oxidation state)

danieleongari commented on aiida_lsmo/utils/cp2k_utils.py:
```diff
@@ -1,53 +1,90 @@
+            'MAGNETIZATION': kind_info['magnetization'],
         })
-        kinds.append({'_': atom + '_ghost', 'BASIS_SET': protocol_settings['basis_set'][atom], 'GHOST': True})
+        kinds.append({
```
This should be named "{kind}_ghost" because the BSSE recognize the tail `_ghost` and uses it for the calculation.

ltalirz replied:
thanks, fixed

danieleongari reviewed (dismissed):
I gave a full read and it seems fine to me.
Some thinking must be done if someone needs to make a DFT calculation after the multistage one, with consistent settings: the wc expose "last_input_settings" but one does now know the atoms KIND correspondence. But it does not look as a big issue now.

ltalirz committed:
consistent example names

ltalirz committed:
reintroduce provenance link

initial magnetization: create new structure in calcfunction
and move call to calcfunction inside workfunction:

AiiDA does not allow calling one calcfunction from another, which forces
us to split the determination of the initial magnetization in a
workfunction and a calcfunction (or loose provenance for oximachine
computation).

Also:
 * Sort Kind symbols generated by initial magnetization guess
   to avoid non-deterministic input file
 * regenerate test data for cp2k calcs

ltalirz commented:
Note: In the last CI run only the coverage upload failed - tests passed.
